web/src/app/modules/actions/modules/export-testspecification-button/components/export-testspecification-button.component.ts

```diff
@@ -0,0 +1,181 @@
+    }
+
+    private pushRows(): void {
+        let testCasesString = '';
+        if (this.testCases) {
+          for (let i = 0 ; i < this.testCases.length ; i++) {
+            testCasesString += this.testCases[i].name + ', ';
+            for (let j = 0 ; j < this.testParameters.length; j++) {
+              let assignmentsList = this.getAssignments(this.testParameters[j]);
+              for (let k = 0 ; k < assignmentsList.length; k++) {
+                if (Url.parent(assignmentsList[k].url) == this.testCases[i].url) {
+                  testCasesString += assignmentsList[k].condition + ', ';
+                  break;
+                }
+              }
+            }
+            testCasesString += '\n';
+          }
+        }
+        this.finalCsvString += testCasesString;
+    }
+    private getAssignments(testParameter: TestParameter): ParameterAssignment[] {
+      let assignmentsList: ParameterAssignment[];
+      assignmentsList = [];
+      for (let i = 0 ; i < this.parameterAssignments.length; i++) {
+        if (this.parameterAssignments[i].parameter.url == testParameter.url) {
+          assignmentsList.push(this.parameterAssignments[i]);
+        }
+      }
+      return assignmentsList;
+    }
+
+    private loadTestCases(): void {
+         this.testCases = this.contents.filter((element: IContainer) => Type.is(element, TestCase))
+             .map((element: IContainer) => element as TestCase);
+    }
+
+    private loadTestParameters(): Promise<void> {
+      let loadTestParametersTask: Promise<void> = Promise.resolve();
+      loadTestParametersTask = loadTestParametersTask.then(() => {
+        this.testParameters = this.contents.filter((element: IContainer) => Type.is(element, TestParameter))
+           .map((element: IContainer) => element as TestParameter);
+      });
+      return loadTestParametersTask;
+    }
+
+    private loadParameterAssignments(): Promise<void>[] {
+        let testCases: TestCase[] = this.testCases;
+        this.parameterAssignments = [];
+        let promiseArray: Promise<void>[];
+        promiseArray = [];
+
+        let loadParameterAssignmentsTask: Promise<void> = Promise.resolve();
+        for (let i = 0; i < testCases.length; i++) {
+            let currentTestCase: TestCase = testCases[i];
+            loadParameterAssignmentsTask = loadParameterAssignmentsTask.then(() => {
+                return this.dataService.readContents(currentTestCase.url)
+                    .then((contents: IContainer[]) =>
+                        contents.forEach((element: IContainer) => {
+                          if (element.className == 'ParameterAssignment') {
+                                this.parameterAssignments.push(element as ParameterAssignment);
+                          }
+                        }));
+            });
+            promiseArray.push(loadParameterAssignmentsTask);
+        }
+        return promiseArray;
+    }
+
+    private createDownloadFile(): void {
```

Comment:
Is there a library that can manage this for us? In my experience, doing all this manually is fragile at least. If there is a library, I am happy to put it into specmate. I think we will need it at some other points as well.

Reply:
@sebeder  can you give me more context. do you mean the way of preparing the string from the loaded arrays (pushRows and pushHeaders) or everything !? you want a library to take the current content array and give us the csv string ?

Reply:
@waleedabdeen, no, I think I know what @sebeder  means. Currently, this is the only place we need this feature, i.e. create a file on the client side such that the browser offers the user to open or save the file. However, we have some use cases in mind where we need that functionality in the future. It would therefore make sense to rely on a tested (on many platforms and browsers) solution that we don't have to maintain ourselves.
Maybe [FileSaver.js](https://github.com/eligrey/FileSaver.js) would work (it has almost as many upvotes as the the solution you chose from SO :P)

Reply:
@munterkalmsteiner  thanks for the clarification, i will look into it and try to implement it.

Reply:
Hi @munterkalmsteiner and @waleedabdeen I think we can need this for the test-case-generation (to allow people to actually download the text generated for test stubs). That's why I'd like to have it as a library.